src/wfs/src/main/java/org/geoserver/wfs/json/GeoJSONBuilder.java

```diff
@@ -139,13 +155,31 @@ private JSONBuilder writeCoordinates(CoordinateSequence coords) throws JSONExcep
 
         // guess the dimension of the coordinate sequence
         int dim = CoordinateSequences.coordinateDimension(coords);
+        // measure
+        int measures = coords.getMeasures();
+        int dimension = coords.getDimension();
 
         final int coordCount = coords.size();
         for (int i = 0; i < coordCount; i++) {
-            if (dim > 2) {
-                writeCoordinate(coords.getX(i), coords.getY(i), coords.getOrdinate(i, 2));
+            // if has not measures coordinate
+            if (measures == 0 || !encodeMeasures) {
+                if (dim > 2) {
+                    writeCoordinate(coords.getX(i), coords.getY(i), coords.getOrdinate(i, 2));
+                } else {
+                    writeCoordinate(coords.getX(i), coords.getY(i));
+                }
             } else {
-                writeCoordinate(coords.getX(i), coords.getY(i));
+                // if is XYZM
+                if (dimension - measures > 2) {
```

Comment:
Quick feedback - doing the logic in the loop is not as fast as doing it outside - and having each block contain its own for lop.

This code is valid but not quite what I expected:

- You have access to coords.hasM() and coords.hasZ() methods if you want to make this logic more readable. If taking this approach consider using the coords.getM(index) and coords.getZ( index ) methods.
- The method used, coords.getOrdinate(index, ordinate), was intended for blindly copying out ordinate values without creating temporary Coordinate object. I kind of wanted it to be fast rather than having to do any logic around dimension and measure.